botcore/utils/cooldown.py

```diff
@@ -0,0 +1,184 @@
+        except TypeError:
+            cooldowns_list = self._cooldowns_non_hashable.get(channel, None)
+            if cooldowns_list is None:
+                return False
+
+            for item in cooldowns_list:
+                if item.call_arguments == call_arguments:
+                    return item.timeout_timestamp > current_time
+            return False
+
+    async def _periodical_cleanup(self, initial_delay: float) -> None:
+        """
+        Wait for `initial_delay`, after that delete stale items every hour.
+
+        The `initial_delay` ensures we're not running cleanups for every command at the same time.
+        """
+        await asyncio.sleep(initial_delay)
+        while True:
+            await asyncio.sleep(60 * 60)
+            self._delete_stale_items()
+
+    def _delete_stale_items(self) -> None:
+        """Remove expired items from internal collections."""
+        current_time = time.monotonic()
+
+        for key, timeout_timestamp in self._cooldowns.copy().items():
+            if timeout_timestamp < current_time:
+                del self._cooldowns[key]
+
+        for key, cooldowns_list in self._cooldowns_non_hashable.copy().items():
+            filtered_cooldowns = [
+                cooldown_item for cooldown_item in cooldowns_list if cooldown_item.timeout_timestamp < current_time
+            ]
+
+            if not filtered_cooldowns:
+                del self._cooldowns_non_hashable[key]
+            else:
+                self._cooldowns_non_hashable[key] = filtered_cooldowns
+
+
+def block_duplicate_invocations(
+    *, cooldown_duration: float = 5, send_notice: bool = False
+) -> Callable[[Callable[P, Awaitable[R]]], Callable[P, Awaitable[R]]]:
+    """
+    Prevent duplicate invocations of a command with the same arguments in a channel for ``cooldown_duration`` seconds.
+
+    Args:
+        cooldown_duration: Length of the cooldown in seconds.
+        send_notice: If True, the user is notified of the cooldown with a reply.
+
+    Returns:
+        A decorator that adds a wrapper which applies the cooldowns.
+
+    Warning:
+        The created wrapper raises :exc:`CommandOnCooldown` when the command is on cooldown.
+    """
+
+    def decorator(func: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]:
+        mgr = _CommandCooldownManager(cooldown_duration=cooldown_duration)
+
+        @command_wraps(func)
+        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+            arg_tuple = (*args[2:], *kwargs.items())
+            ctx = typing.cast("Context[BotBase]", args[1])
+            channel = ctx.channel
+
+            if not isinstance(channel, discord.DMChannel):
+                if mgr.is_on_cooldown(ctx.channel, arg_tuple):
+                    if send_notice:
+                        with suppress(discord.NotFound):
+                            await ctx.reply("The command is on cooldown with the given arguments.")
```

Comment:
Should the time remaining be included in the message?

Reply:
Yeah, fair point. I don't anticipate this to be used with long cooldowns. We can revisit it at any point in the future if we change our minds. It's a simple change.